src/claim_proofs/ffi_wrapper/src/lib.rs

```diff
@@ -0,0 +1,271 @@
+//! The `claim_proofs_ffi` is the Foreign Function Interface (FFI)
+//! for the `claim_proofs` library. It contains API for generating
+//! claim proofs and verifying them as part of the
+//! Asset Granularity Unique Identity project.
+
+extern crate libc;
+use libc::size_t;
+use std::slice;
+
+use cryptography::claim_proofs::{build_scope_claim_proof_data, compute_cdd_id, compute_scope_id};
+
+pub type ScopeClaimData = cryptography::claim_proofs::ScopeClaimData;
+pub type ScopeClaimProofData = cryptography::claim_proofs::ScopeClaimProofData;
+pub type ProofPublicKey = cryptography::claim_proofs::ProofPublicKey;
+pub type ProofKeyPair = cryptography::claim_proofs::ProofKeyPair;
+pub type CDDClaimData = cryptography::claim_proofs::CDDClaimData;
+pub type Signature = schnorrkel::Signature;
+pub type Scalar = curve25519_dalek::scalar::Scalar;
+pub type RistrettoPoint = curve25519_dalek::ristretto::RistrettoPoint;
+
+fn box_alloc<T>(x: T) -> *mut T {
+    Box::into_raw(Box::new(x))
+}
+
+// ------------------------------------------------------------------------
+// Data Structures
+// ------------------------------------------------------------------------
+
+/// Create a `Scalar` object from a 32 bytes array.
+///
+/// Caller is responsible for calling `scalar_free()` to deallocate this object.
+/// SAFETY: Caller is also responsible for making sure the `scalar_bits` pointer
+///         is not Null and points to a 32 bytes block of memory.
+#[no_mangle]
+pub unsafe extern "C" fn scalar_new(scalar_bits: *const u8) -> *mut Scalar {
+    assert!(!scalar_bits.is_null());
```

Comment:
Per previous discussion, this `assert!` and the `if ptr.is_null() { return; }` below should be removed (applies throughout)?

Reply:
The `if ptr.is_null() { return; }` is C convention for `free()` functions. It avoids attempting to double-free a pointer. I can't remove that.

What's the harm with keeping the assert? it will panic if the caller calls with a Null pointer.

Reply:
Well the `assert!` and the `if ptr....` does pessimize performance, but if that is OK with you, then keep it.

Reply:
I have 8 years of experience developing crypto libraries in C, I might not be the most proficient in Rust though ;), I think it's safer to keep these, even at the cost of performance. :)

Reply:
Fair enough. :)